test/integration/core/hash-consensus.integration.ts

```diff
@@ -0,0 +1,252 @@
+
+    // Get slots per epoch from chain config
+    const [slotsPerEpoch] = await hashConsensus.getChainConfig();
+
+    // Test epochs per frame cannot be zero
+    await expect(hashConsensus.connect(stranger).setFrameConfig(0, 1)).to.be.revertedWithCustomError(
+      hashConsensus,
+      "EpochsPerFrameCannotBeZero",
+    );
+
+    // Test fast lane period cannot be longer than frame
+    const epochsPerFrame = 2n;
+    await expect(
+      hashConsensus.connect(stranger).setFrameConfig(epochsPerFrame, epochsPerFrame * slotsPerEpoch + 1n),
+    ).to.be.revertedWithCustomError(hashConsensus, "FastLanePeriodCannotBeLongerThanFrame");
+  });
+
+  it("Should validate fast lane length slots parameter", async () => {
+    // Grant MANAGE_FAST_LANE_CONFIG_ROLE to stranger
+    const MANAGE_FAST_LANE_CONFIG_ROLE = await hashConsensus.MANAGE_FAST_LANE_CONFIG_ROLE();
+    await hashConsensus.connect(agent).grantRole(MANAGE_FAST_LANE_CONFIG_ROLE, stranger.address);
+
+    // Get chain config and frame config
+    const [slotsPerEpoch] = await hashConsensus.getChainConfig();
+    const [, epochsPerFrame] = await hashConsensus.getFrameConfig();
+
+    // Test fast lane period cannot be longer than frame
+    await expect(
+      hashConsensus.connect(stranger).setFastLaneLengthSlots(epochsPerFrame * slotsPerEpoch + 1n),
+    ).to.be.revertedWithCustomError(hashConsensus, "FastLanePeriodCannotBeLongerThanFrame");
+  });
+
+  it("Should validate member addition", async () => {
+    // Grant MANAGE_MEMBERS_AND_QUORUM_ROLE to stranger
+    const MANAGE_MEMBERS_AND_QUORUM_ROLE = await hashConsensus.MANAGE_MEMBERS_AND_QUORUM_ROLE();
+    await hashConsensus.connect(agent).grantRole(MANAGE_MEMBERS_AND_QUORUM_ROLE, stranger.address);
+
+    const currentQuorum = await hashConsensus.getQuorum();
+    const [members] = await hashConsensus.getMembers();
+
+    // Test cannot add duplicate member
+    await expect(hashConsensus.connect(stranger).addMember(members[0], currentQuorum)).to.be.revertedWithCustomError(
+      hashConsensus,
+      "DuplicateMember",
+    );
+
+    // Test cannot add zero address
+    await expect(hashConsensus.connect(stranger).addMember(ZeroAddress, currentQuorum)).to.be.revertedWithCustomError(
+      hashConsensus,
+      "AddressCannotBeZero",
+    );
+
+    // Test quorum must be greater than half members + 1
+    const requiredQuorum = Math.floor((members.length + 1) / 2) + 1;
+    await expect(hashConsensus.connect(stranger).addMember(stranger.address, requiredQuorum - 1))
+      .to.be.revertedWithCustomError(hashConsensus, "QuorumTooSmall")
+      .withArgs(requiredQuorum, requiredQuorum - 1);
+  });
+
+  it("Should validate member removal", async () => {
+    // Grant MANAGE_MEMBERS_AND_QUORUM_ROLE to stranger
+    const MANAGE_MEMBERS_AND_QUORUM_ROLE = await hashConsensus.MANAGE_MEMBERS_AND_QUORUM_ROLE();
+    await hashConsensus.connect(agent).grantRole(MANAGE_MEMBERS_AND_QUORUM_ROLE, stranger.address);
```

Comment:
Already granted? Because no snapshot reset between the tests

Reply:
no longer relevant, because tests have been isolated